File: sdk/typescript/src/roots.ts
import { mkdirSync } from "node:fs";
import { join, resolve } from "node:path";
import type { RuntimeRootsOptions, RuntimeSkillRoot } from "./types.js";

/**
 * Helper utilities for building the formal ROOT, PROJECT, USER skill-root chain.
 * 用于构造正式 ROOT、PROJECT、USER 技能根链的辅助工具。
 */
export class RuntimeRoots {
  /**
   * Build one standard formal root chain from a shared runtime root.
   * 基于共享 runtime root 构造一条标准正式根链。
   */
  static standard(options: RuntimeRootsOptions | string): RuntimeSkillRoot[] {
    const normalizedOptions = typeof options === "string" ? { runtimeRoot: options } : options;
    const runtimeRoot = resolve(normalizedOptions.runtimeRoot);
    const roots: RuntimeSkillRoot[] = [
      {
        name: "ROOT",
        skills_dir: join(runtimeRoot, normalizedOptions.rootSkillsDirName ?? "root_skills"),
      },
    ];
    if (normalizedOptions.includeProject ?? true) {
      roots.push({
        name: "PROJECT",
        skills_dir: join(runtimeRoot, normalizedOptions.projectSkillsDirName ?? "project_skills"),
      });
    }
    if (normalizedOptions.includeUser ?? true) {
      roots.push({
        name: "USER",
        skills_dir: join(runtimeRoot, normalizedOptions.userSkillsDirName ?? "user_skills"),
      });
    }
    return roots;
  }

  /**
   * Build one ROOT-only chain for hosts that intentionally expose no ordinary skill layer.
   * 为明确不暴露普通 skill 层的宿主构造一条仅 ROOT 的根链。
   */
  static rootOnly(runtimeRoot: string): RuntimeSkillRoot[] {
    return RuntimeRoots.standard({
      runtimeRoot,
      includeProject: false,
      includeUser: false,
    });
  }

  /**
   * Find one root by formal label using the same trim and uppercase convention as the runtime.
   * 按运行时相同的 trim 与大写约定查找单个正式标签 root。
   */
  static findByLabel(skillRoots: RuntimeSkillRoot[], label: "ROOT" | "PROJECT" | "USER" | string): RuntimeSkillRoot {
    const normalizedLabel = RuntimeRoots.normalizeLabel(label);
    const root = skillRoots.find((candidate) => RuntimeRoots.normalizeLabel(candidate.name) === normalizedLabel);
    if (!root) {
      throw new Error(`Runtime skill root '${normalizedLabel}' is not present in the configured root chain`);
    }
    return root;
  }

  /**
   * Create runtime directories needed by the default SDK host options and root chain.
   * 创建默认 SDK 宿主选项和 root 链所需的运行时目录。
   */
  static ensureLayout(runtimeRoot: string, skillRoots: RuntimeSkillRoot[] = RuntimeRoots.standard(runtimeRoot)): void {
    const absoluteRoot = resolve(runtimeRoot);
    const directories = [
      absoluteRoot,
      join(absoluteRoot, "temp"),
      join(absoluteRoot, "temp", "downloads"),
      join(absoluteRoot, "resources"),
      join(absoluteRoot, "lua_packages"),
      join(absoluteRoot, "bin", "tools"),
      join(absoluteRoot, "libs"),
      join(absoluteRoot, "dependencies"),
      join(absoluteRoot, "state"),
      join(absoluteRoot, "databases"),
      ...skillRoots.map((root) => root.skills_dir),
    ];
    for (const directory of directories) {
      mkdirSync(directory, { recursive: true });
    }
  }

  /**
   * Normalize one formal root label for client-side lookup.
   * 为客户端侧查找归一化单个正式 root 标签。
   */
  private static normalizeLabel(label: string): string {
    return label.trim().toUpperCase();
  }
}

File: sdk/go/roots.go
package luaskills

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"
)

// StandardRoots builds one formal ROOT, PROJECT, USER root chain from a shared runtime root.
// StandardRoots 基于共享 runtime root 构造一条正式 ROOT、PROJECT、USER root 链。
func StandardRoots(runtimeRoot string) []RuntimeSkillRoot {
	return StandardRootsWithOptions(runtimeRoot, true, true)
}

// StandardRootsWithOptions builds one formal root chain with optional PROJECT and USER layers.
// StandardRootsWithOptions 构造一条可选 PROJECT 与 USER 层的正式 root 链。
func StandardRootsWithOptions(runtimeRoot string, includeProject bool, includeUser bool) []RuntimeSkillRoot {
	root := normalizePath(runtimeRoot)
	roots := []RuntimeSkillRoot{{
		Name:      "ROOT",
		SkillsDir: normalizePath(filepath.Join(root, "root_skills")),
	}}
	if includeProject {
		roots = append(roots, RuntimeSkillRoot{
			Name:      "PROJECT",
			SkillsDir: normalizePath(filepath.Join(root, "project_skills")),
		})
	}
	if includeUser {
		roots = append(roots, RuntimeSkillRoot{
			Name:      "USER",
			SkillsDir: normalizePath(filepath.Join(root, "user_skills")),
		})
	}
	return roots
}

// RootOnly builds one ROOT-only chain for system-only hosts.
// RootOnly 为仅系统层宿主构造一条仅 ROOT 的 root 链。
func RootOnly(runtimeRoot string) []RuntimeSkillRoot {
	return StandardRootsWithOptions(runtimeRoot, false, false)
}

// FindRootByLabel locates one root by formal label using trim and uppercase normalization.
// FindRootByLabel 使用 trim 与大写归一化按正式标签定位单个 root。
func FindRootByLabel(skillRoots []RuntimeSkillRoot, label string) (*RuntimeSkillRoot, error) {
	normalizedLabel := strings.ToUpper(strings.TrimSpace(label))
	for index := range skillRoots {
		if strings.ToUpper(strings.TrimSpace(skillRoots[index].Name)) == normalizedLabel {
			return &skillRoots[index], nil
		}
	}
	return nil, fmt.Errorf("runtime skill root %q is not present in the configured root chain", normalizedLabel)
}

// EnsureRuntimeLayout creates runtime directories needed by default SDK host options and root chain.
// EnsureRuntimeLayout 创建默认 SDK 宿主选项和 root 链所需的运行时目录。
func EnsureRuntimeLayout(runtimeRoot string, skillRoots []RuntimeSkillRoot) error {
	root := normalizePath(runtimeRoot)
	if len(skillRoots) == 0 {
		skillRoots = StandardRoots(root)
	}
	directories := []string{
		root,
		filepath.Join(root, "temp"),
		filepath.Join(root, "temp", "downloads"),
		filepath.Join(root, "resources"),
		filepath.Join(root, "lua_packages"),
		filepath.Join(root, "bin", "tools"),
		filepath.Join(root, "libs"),
		filepath.Join(root, "dependencies"),
		filepath.Join(root, "state"),
		filepath.Join(root, "databases"),
	}
	for _, skillRoot := range skillRoots {
		directories = append(directories, filepath.FromSlash(skillRoot.SkillsDir))
	}
	for _, directory := range directories {
		if err := os.MkdirAll(directory, 0o755); err != nil {
			return err
		}
	}
	return nil
}

// normalizePath returns one absolute slash-normalized filesystem path.
// normalizePath 返回单个绝对且斜杠归一化的文件系统路径。
func normalizePath(path string) string {
	absolute, err := filepath.Abs(path)
	if err != nil {
		return filepath.ToSlash(path)
	}
	return filepath.ToSlash(absolute)
}
